### PR TITLE
fix(Transaction): discard empty transactions on CommitWith 

### PR DESCRIPTION
**Pull request by Wondertan**
## Problems
* Transactions with empty `pendingWrites` were never discarded(and marked as done) for `CommitWith`
	* This is similar to https://github.com/dgraph-io/badger/pull/2018, which solved the same problem for `Commit`, but not for `CommitWith`
	* The `CommitWith` is used by `WriteBatch`, so flushing an empty batch never discarded the inner shadowed transaction, causing a multitude of issues.

## Solution
Make sure `Discard` is called for `CommitWith` when `pendingWrites` are empty. The existing unit test is updated to assert that.

### CODE REVIEW

**Comment by Wondertan** [comment]
cc @billprovince @joshua-goldstein

File: txn.go [hunk 1]
```diff
@@ -712,6 +711,7 @@ func (txn *Txn) CommitWith(cb func(error)) {
 	if cb == nil {
 		panic("Nil callback provided to CommitWith")
 	}
+	defer txn.Discard()
```
Comment (harshil-goel) [inline txn.go:714]: Why two defer for the sam ething?

Comment (Wondertan) [reply txn.go:714]: I don't understand the question

Comment (Wondertan) [reply txn.go:714]: Could you rephrase it?

Comment (Wondertan) [reply txn.go:714]: Yeah, I get it, but what's the same thing? If that's about `Commit` and `CommitWith`, then it's not the same. They are two different ways of committing data and both have to Discard.

Also note that Discard protection from multiple calls, and even if a user calls both there won't be any side effects.

File: txn_test.go [hunk 2]
```diff
@@ -761,7 +761,6 @@ func TestManagedDB(t *testing.T) {
 		for i := 0; i <= 3; i++ {
 			require.NoError(t, txn.SetEntry(NewEntry(key(i), val(i))))
 		}
-		require.Error(t, txn.Commit())
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply txn_test.go:764]: To avoid committing twice.

[the remark this answers is not in the document]
Comment (Wondertan) [reply txn_test.go]: There was no good reason, so I took it back to the original state.

**Comment by Wondertan** [comment]
Ok, so I think I misunderstood the purpose of `precommitCheck`, and transactions should not be discarded when it errors. I still think the code in Commit funcs is rather confusing and could be simplified to have a single defer, but that's definitely not part of this PR, so I shortened the solution to a one-liner, as well updated unit tests that checks for the bug

**Comment by Wondertan** [comment]
Ping again

**Review by mangalaman93 — APPROVED** [review]
